### ISSUE TITLE
jax.numpy.concatenate is much slower than np.concatenate

### ISSUE THREAD

**Issue by jon-deng** [google/jax#4865]
Hi,

While using JAX I noticed there is a large slowdown when using `jax.numpy.concatenate`. A minimum example is:

```python
from time import perf_counter

import numpy as np
import jax.numpy as jnp

def f(x, y):
    return np.concatenate((x, y))

def g(x, y):
    return jnp.concatenate((x, y))

if __name__ == '__main__':
    NN = 10000
    x, y = np.ones(5), np.ones(6)

    time_f = 0
    time_g = 0
    for n in range(NN):
        t0 = perf_counter()
        _ = f(x, y)
        time_f += perf_counter() - t0

        t0 = perf_counter()
        _ = g(x, y)
        time_g += perf_counter() - t0

    print(f"Average time for numpy {time_f/NN/1e-6:.4f} us")
    print(f"Average time for jax {time_g/NN/1e-6:.4f} us")
```

On my system I'm getting 6 us and 137 us for numpy and jax, respectively. As a result, my application is slowed because concatenate is taking most of the time.

Any ideas why this happens or how to fix this? I'm using JAX 0.2.5 from conda-forge. I also tried a work around with `jax.ops.index_update` but it doesn't seem to improve the speed at all.

Thanks for the help ahead of time!

**Comment by mattjj** [comment]
Thanks for the question!

Actually, due to [JAX's async dispatch](https://jax.readthedocs.io/en/latest/async_dispatch.html), this timing isn't quite what you want, and the actual results could be even _worse_ for JAX (though they weren't in my experiment below):

```python
from time import perf_counter

import numpy as np
import jax.numpy as jnp

def f(x, y):
    return np.concatenate((x, y))

def g(x, y):
    return jnp.concatenate((x, y))

if __name__ == '__main__':
    NN = 10000
    x, y = np.ones(5), np.ones(6)

    time_f = 0
    time_g = 0
    for n in range(NN):
        t0 = perf_counter()
        _ = f(x, y)
        time_f += perf_counter() - t0

        t0 = perf_counter()
        _ = g(x, y).block_until_ready()  # note block_until_ready()
        time_g += perf_counter() - t0

    print(f"Average time for numpy {time_f/NN/1e-6:.4f} us")
    print(f"Average time for jax {time_g/NN/1e-6:.4f} us")
```

```
$ python 4865.py
WARNING:absl:No GPU/TPU found, falling back to CPU. (Set TF_CPP_MIN_LOG_LEVEL=0 and rerun for more info.)
Average time for numpy 2.9865 us
Average time for jax 80.5339 us
```

Most of what's being timed here is JAX's dispatch time. That is, we're mostly looking at how long it takes JAX to start executing the actual concatenation operation, due to Python overheads as well as lower-level runtime overheads. The usual ballpark is that JAX dispatch overheads can be ~100µs, whereas NumPy's are often <1µs. (PyTorch's overheads are close to NumPy's, like maybe 1-2µs, though AIUI to dispatch a GPU kernel can take 5-10µs anyway.)

Those overheads can be optimized, and there's some early work starting to do that now, but in general we haven't gotten around to optimizing these kinds of dispatch overheads because most users rely on `jit` for performance. Under a `jit`, these kinds of overheads don't exist, so as long as you can `jit` big enough functions (say, a neural network training step that takes at least milliseconds to execute), these kinds of dispatch overheads aren't very important.

However, there are still workloads where fast dispatch is important, perhaps because we can't `jit` big enough functions. In those cases, NumPy (and PyTorch) have a big advantage on dispatch times.

Can you use `jit` in your application? If not, it may be that JAX isn't the right tool, at least until we optimize dispatch overheads. (I'm optimistic that sometime in this calendar year we'll get dispatch overheads for computations like this one under 5µs.)

WDYT?

**Comment by mattjj** [comment]
If you upgrade to jaxlib==0.1.57, which we just pushed to pypi a few minutes ago, then the time is actually much better using `jit`:

```python
from time import perf_counter

import numpy as np
import jax.numpy as jnp
from jax import jit

def f(x, y):
    return np.concatenate((x, y))

@jit  # notice this new jit call!          <--------
def g(x, y):
    return jnp.concatenate((x, y))

if __name__ == '__main__':
    NN = 10000
    x, y = np.ones(5), np.ones(6)

    time_f = 0
    time_g = 0
    for n in range(NN):
        t0 = perf_counter()
        _ = f(x, y)
        time_f += perf_counter() - t0

        t0 = perf_counter()
        _ = g(x, y).block_until_ready()  # note block_until_ready()
        time_g += perf_counter() - t0

    print(f"Average time for numpy {time_f/NN/1e-6:.4f} us")
    print(f"Average time for jax {time_g/NN/1e-6:.4f} us")
```

```
Average time for numpy 2.5906 us
Average time for jax 40.2877 us
```

On jaxlib 0.1.56 it was 80µs for me yesterday, and now it's 40µs today! If we keep this up we'll be at 1e-100µs in no time.

**Comment by jon-deng** [comment]
Thanks so much for the help! I understand now, each `jnp.function` call has a relatively high overhead for dispatching so that's why it takes longer than numpy for a single operation (I should have read the documentation more).

In my application the function has a bunch more operations other than just `jnp.concatenate` so using `jit` will definitely improve the speed. Upgrading to the latest version should help even more. Thanks again!

**Comment by mattjj** [comment]
Glad it helped!

**Comment by swamidass** [comment]
Is concatenate (on the same device) a no copy operation?

**Comment by jakevdp** [comment]
No, in general concatenation requires copying the inputs into a new buffer, because the array model is not flexible enough to treat unstructured memory chunks as a single array. This is the case in both JAX and NumPy.
